akashic-cli-serveに起動画面表示オプションと起動画面を追加

dera- opened the pull request:
### 概要
* akashic-cli-serveからコンテンツに対して引数を渡す機能が無かったので今回追加しました。
  * 引数を渡す場合は、-A(--no-auto-start)オプションを付けてakashic-cli-serveを起動して、コンテンツを開くと表示される起動画面上のUIから送ることになります

### やったこと
* 起動画面を表示するかどうかのコマンドオプション(-A, --no-auto-start)を追加
* クライアント側で以下のことができる起動画面を表示できるようにした
  * コンテンツにjoinするかどうかの選択
  * コンテンツに送る引数の編集
  * 起動引数候補リストの表示 (copyボタンでtextareaにコピー)
    * 起動引数候補リストの内容は、sandbox.config.js でゲーム開発者が指定できる
  * ゲーム実行開始ボタン
* 以下の内部APIを追加
  * /config/options: サーバー起動時の各種オプションの値を取得
* 以下の内部APIの仕様を変更
  * /api/sandbox-config: 起動引数候補リストを値とする`arguments`キーを追加した

### その他
* UIについては別PRで改修予定です

dera- committed:
add startup-window to akashic-cli-serve

dera- commented on packages/akashic-cli-serve/src/client/operator/Operator.ts:
```diff
@@ -62,16 +68,32 @@ export class Operator {
+		}
+	}
+
+	startContent = async (params?: StartContentParameterObject): Promise<void> => {
```
起動画面の追加に伴って、playの生成とLocalInstanceの生成を分けることにしました

dera- commented on packages/akashic-cli-serve/src/client/store/StartupScreenUiStore.ts:
```diff
@@ -0,0 +1,35 @@
+	@observable startupArgumentEditContent: string;
+	@observable joinFlag: boolean;
+
+	constructor() {
```
起動画面のUIの各種値は引き回すことはないと思われますので、storageは利用しない感じにしました

yu-ogi replied:
個人的には保存されてほしいなと思っています。

dera- replied:
なるほど。。ちょっと検討してみます

dera- replied:
こちらstorageを利用するように変更してみました

dera- commented on packages/akashic-cli-serve/src/client/store/StartupScreenUiStore.ts:
```diff
@@ -0,0 +1,35 @@
+
+	constructor() {
+		this.showsStartupArgumentList = true;
+		this.startupArgumentListWidth = 150;
```
手抜きでマジックナンバーにしてしまいました。。どこかにUI用の定数を定義できるところがあればいいのですが。。

dera- commented on packages/akashic-cli-serve/src/client/view/atom/RightResizable.tsx:
```diff
@@ -35,7 +35,6 @@ export class RightResizable extends React.Component<RightResizableProps, {}> {
 	private _onMouseMoveWindow = (ev: MouseEvent): void => {
 		const { onResize, width, minWidth } = this.props;
 		const w = Math.max(width + (ev.pageX - this.lastPageX), minWidth);
-		console.log({ py: ev.pageX, sy: this.lastPageX, diff: (ev.pageX - this.lastPageX) });
```
これらのconsoleは不要かと思ったので削除しました。削除していけないものであれば教えていただきたいです

xnv replied:
消す忘れですね……ありがとうございます。

dera- commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.css:
```diff
@@ -0,0 +1,136 @@
+.devtool {
```
このcssの内容は、後から修正していただく前提で用意したので、現状はほとんどEventsDevtool.cssとDevtool.cssの内容のコピーとなっています。

yu-ogi replied:
将来的には scss あたりで共通化したいですね。

dera- commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.tsx:
```diff
@@ -0,0 +1,99 @@
+import * as React from "react";
```
起動画面コンポーネントを暫定的にこのファイル1つでまとめました。
コンポーネントを切り出す余地はありそうですが、その対応はやるとしたら別PRになるかと思います

xnv commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.tsx:
```diff
@@ -0,0 +1,99 @@
+	onClickCopyEvent: (eventName: string) => void;
+	onEventEditContentChanged: (content: string) => void;
+	onChangeJoinFlag: (joinFlag: boolean) => void;
+	startContent: (params?: StartContentParameterObject) => Promise<void>;
```
他と名前を合わせて欲しいです。 onClickStartContent とかでしょうか。

dera- replied:
onClickStartContent に修正しました

xnv commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.tsx:
```diff
@@ -0,0 +1,99 @@
+	onToggleList: (nextVal: boolean) => void;
+	eventNames: string[];
+	eventEditContent: string;
+	joinFlag: boolean;
```
これも。 `showsEventList` い倣うなら `joinsToPlay` とかでしょうか。

dera- replied:
了解しました、joinFlag使っている箇所をjoinsToPlayに変更しました

xnv commented on packages/akashic-cli-serve/src/client/api/ApiClient.ts:
```diff
@@ -7,7 +7,7 @@ import {
 	RunnerPostApiResponse,
 	RunnerPatchApiResponse,
-	SandboxConfigApiResponse
+	SandboxConfigApiResponse, OptionsApiResponse, StartupArgumentApiResponse
```
周辺と合わせて改行をお願いしたいですね……。

dera- replied:
こちら修正しました

xnv commented on packages/akashic-cli-serve/src/client/bootstrap.tsx:
```diff
@@ -16,11 +16,13 @@ const operator = new Operator({
+	try {
+		await operator.bootstrap();
+	} catch (e) {
+		console.error(e);
```
これはどういうケースの想定なのでしょうか。 bootstrap() できない状況で render() する……？

dera- replied:
こちらは以前に書かれていたものを踏襲した感じですね。ちなみにこの変更はload時のみbootstrapが実行されるようにしたかったので行いました。

xnv commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.tsx:
```diff
@@ -0,0 +1,99 @@
+				<div className={styles["events-toolbar"]}>
+					<ToolIconButton
+						icon="list"
+						title={"イベントリストの表示・非表示を切り替え"}
```
起動引数リストの〜にしたいです。 (後からいじるとはいえ tooltip に事実と異なる表示が出ちゃうので)

dera- replied:
修正しました

xnv commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.tsx:
```diff
@@ -0,0 +1,99 @@
+				<textarea
+					className={styles["editor"]}
+					value={props.eventEditContent}
+					placeholder={"an array of playlog events (JSON) to send"}
```
ここもイベントのままになっています。

dera- replied:
こちらも修正しました

xnv commented on packages/akashic-cli-serve/src/server/controller/StartupArgumentController.ts:
```diff
@@ -0,0 +1,27 @@
+			};
+			try {
+				// TODO ファイル監視。内容に変化がなければ直前の値を返せばよい
+				data.args = require(path.resolve(dirpath, "startup_argument.json"));
```
イベントリスト同様に sandbox.config.js の1フィールドでいいかと思います。 "events" に倣って "arguments" でどうでしょうか。

dera- replied:
sandbox.config.jsを使うように修正しました。同時にSandboxConfig型にもarguments要素を追加しました

dera- committed:
reflect comment

dera- committed:
fix lerna.json temporarily

Commit:
Publish

 - @akashic/akashic-cli-serve@0.0.10-alpha.0
 - @akashic/akashic-cli@1.5.3-alpha.0

dera- committed:
resolve conflict when merged with master

Commit:
Publish

 - @akashic/akashic-cli-serve@0.0.11-alpha.0
 - @akashic/akashic-cli@1.5.4-alpha.0

yu-ogi commented on packages/akashic-cli-serve/src/server/route/ConfigRoute.ts:
```diff
@@ -8,6 +8,7 @@ export const createConfigRouter = (): express.Router => {
 	// /engineとの相違点はスクリプトアセット加工前のコンテンツを含む情報を投げること
 	// サーバー側でインスタンスを立ち上げる時は加工前のスクリプトアセットを参照する必要がある
 	configRouter.get("/content.raw.json", createHandlerToGetEngineConfig(true));
+	configRouter.get("/options", handleToGetStartupOptions);
```
`create...` と `handle...` の命名の違いの意図って何でしょう?
理由がなければ揃えたほうが良さそうに思いました。

dera- replied:
`create...` はハンドラ関数を生成しているもので、`handle...` はハンドラ関数自体という区別がありますね。
そもそも `handle...` を定義するのではなくハンドラ関数を生成する `create...` で定義するに揃えたらいいのではとも思われそうですが、外部から渡されるものもなく常に作られるハンドラは同じになるものについてはパフォーマンス的なことも考慮して(それほど影響はなさそうですが)ハンドラ関数の定義だけでいいのではと思ってこのようにしてます

yu-ogi replied:
なるほど、見落としていました。

yu-ogi commented on packages/akashic-cli-serve/src/server/index.ts:
```diff
@@ -37,7 +37,8 @@ export function run(argv: any): void {
 		.option("-H, --hostname <hostname>", `The host name of the server. default: ${serverGlobalConfig.hostname}`)
-		.option("-v, --verbose", `Display detailed information on console.`)
+		.option("-v, --verbose", `Display detailed information on console. default: false`)
+		.option("-w, --wait", `Stop automatic startup of contents. default: false`)
```
- 説明が `Stop ...` で始まるなら `--stop` のほうが適切に思えました。
   - `--wait` であるなら `Wait ..` で説明を始めたほうが良さそうに思えます。
- `-w` って必要ですかね?

dera- replied:
> `--wait` であるなら `Wait ..` で説明を始めたほうが良さそうに思えます。

確かにそうですね。`Wait ..` で説明を書いてみます。

> `-w` って必要ですかね?

すみません、よく分かっていないのですが、これは `wait` オプション自体必要ないのではということでしょうか？であれば、起動画面の表示を省略したい場合もあると思うので僕は必要だと思ってますが。。

yu-ogi replied:
> すみません、よく分かっていないのですが、これは `wait` オプション自体必要ないのではということでしょうか？

いえ、省略した引数 `-w` のことです。

dera- replied:
うーん、省略形はあってもいいと思うんですよね。。個人的な話ですみませんが、僕としては省略系の方を多用しますし。
逆にあるとまずかったりしますでしょうか？

yu-ogi replied:
> 逆にあるとまずかったりしますでしょうか？

別にまずくないと思います。

- 今後 `--w...` で始まる引数が追加された場合に `-w` が使えなくなる
- そもそも自動で開始される動作こそ引数が必要 ( `--auto-start` など) なのではないか

など考えられる中で `-w` を今すぐ使ってしまう必要性があるのか?という意味で質問いたしました。

dera- replied:
なるほど。ちょっとその件については後日相談させてください。
現状プレリリースバージョンで `-w` が使える状態になってしまっているため、いきなり無くすことを即決できない状況ですので。。

dera- replied:
別で相談したところ、`--wait`オプション自体暫定的なものなので一旦`--wait`だけにするという結論になりました。
また`--wait`オプションの名前をどうするかやそもそも自動開始の方をオプションにすべきだとかそういった議論は別途相談ということになりました

dera- replied:
auto-startの方がデフォルトということなので、デフォルトとは違う挙動をするというのを直感的にするために--no-auto-start というオプション名に変えてみました

yu-ogi commented on packages/akashic-cli-serve/src/client/store/LocalInstanceEntity.ts:
```diff
@@ -26,6 +26,7 @@ export interface LocalInstanceEntityParameterObject {
 	executionMode: ExecutionMode;
 	play: PlayEntity;
 	player: Player;
+	startupArgument?: any;
```
(imo) `LocalInstanceEntity` からみたら `startup` なのは自明なので `argument` で良いと思います。

dera- replied:
修正しておきます

yu-ogi commented:
PullRequst の説明にある

> * /api/atartup-argument: 対象コンテンツのディレクトリ上にあるstartup_argument.jsonの内容を取
得

はどこに処理があるのでしょう?

dera- commented:
> * /api/atartup-argument: 対象コンテンツのディレクトリ上にあるstartup_argument.jsonの内容を取
  得

このAPIですが使わないことにして削除したので、上記説明を修正しておきます。
ちなみに、代わりに起動引数候補リストは/api/sandbox-configから取得できるようにしています

dera- committed:
reflect comment

dera- committed:
resolve conflict when merged with master

yu-ogi commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.tsx:
```diff
@@ -0,0 +1,100 @@
+}
+
+@observer
+export class StartupScreen extends React.Component<StartupScreenProps, {}> {
```
`Screen` というものが初めて使われていますが `Screen` とはどのようなものでしょう?

dera- replied:
「起動画面」の「画面」という意味で使いたかった(新しい概念を追加したかったわけでなく単なる名詞として入れたいという意図)のですが、screenはあまり一般的なものではない感じでしょうか？

yu-ogi commented on packages/akashic-cli-serve/src/client/view/organism/StartupScreen.tsx:
```diff
@@ -0,0 +1,100 @@
+		</div>;
+	}
+
+	private _generateStartArgumentDom = (): React.ReactNode => {
```
場所によって `startupArgument` だったり `startArgument` だったりと異なる名前になってしまっているので、できれば統一したいです。

dera- replied:
お、それはケアレスミスな気がしますね。。僕としては`startupArgument`で統一したい気持ちなので`startArgument`となっているところを直そうと思います。

yu-ogi commented on packages/akashic-cli-serve/src/client/store/storage.ts:
```diff
@@ -9,6 +9,10 @@ export interface StorageData {
+	showsStartupArgumentList: boolean;
+	startupArgumentListWidth: number;
+	startupArgumentEditContent: string;
+	joinsToPlay: boolean;
```
自動的に `join` する、という意なら `joinsAuto(matically)` などの方が適切かなと思いました。

dera- replied:
なるほど。確かに意味としては「ゲーム開始時に自動でjoinする」なのでそちらの方が適切に思えます。修正しておきます。

dera- committed:
reflect comment

dera- committed:
resolve conflict when merged with master

dera- committed:
rename wait option to no-auto-start option

yu-ogi commented on packages/akashic-cli-serve/src/server/index.ts:
```diff
@@ -83,6 +86,10 @@ export function run(argv: any): void {
 	}
 
+	if (commander.autoStart != null) {
+		serverGlobalConfig.waitStarting = !commander.autoStart;
```
特に理由がなければ他と合わせて `serverGlobalConfig` 側も `autoStart` にしたいです。

dera- replied:
修正します

dera- committed:
reflect comment

xnv reviewed (approved):
起動引数/startup argument なのだと思いますが、どちらかというと「ゲームコンテンツの引数」であることの方が強調したい気もし、 game argument とか content argument とかはどうでしょうか。

……というコメントはありますが approved です。

dera- committed:
rename startupArgument to gameArgument
